tarantool/connection_pool.py

```diff
@@ -0,0 +1,406 @@
+            self.pool)
+
+    def _request_process_loop(self, key, unit, last_refresh):
+        while unit.request_process_enabled:
+            if not unit.input_queue.empty():
+                task = unit.input_queue.get()
+                method = getattr(Connection, task.method_name)
+                try:
+                    resp = method(unit.conn, *task.args, **task.kwargs)
+                except Exception as e:
+                    unit.output_queue.put(e)
+                else:
+                    unit.output_queue.put(resp)
+
+            now = time.time()
+
+            if now - last_refresh > self.refresh_delay:
+                self._refresh_state(key)
+                last_refresh = time.time()
+
+    def connect(self):
+        for key in self.pool:
+            unit = self.pool[key]
+
+            self._refresh_state(key)
+            last_refresh = time.time()
+
+            unit.thread = threading.Thread(
+                target=self._request_process_loop,
+                args=(key, unit, last_refresh),
+                daemon=True,
+            )
+            unit.request_process_enabled = True
+            unit.thread.start()
+
+    def _send(self, mode, method_name, *args, **kwargs):
+        key = self.strategy.getnext(mode)
+        unit = self.pool[key]
+
+        task = PoolTask(method_name=method_name, args=args, kwargs=kwargs)
+
+        unit.input_queue.put(task)
+        resp = unit.output_queue.get()
+
+        if isinstance(resp, Exception):
+            raise resp
+
+        return resp
+
+    def call(self, func_name, *args, mode=Mode.PREFER_RW):
+        '''
+        :param tarantool.Mode mode: Request mode (default is
+        PREFER_RW).
+        '''
+        return self._send(mode, 'call', func_name, *args)
+
+    def eval(self, expr, *args, mode=Mode.PREFER_RW):
+        '''
+        :param tarantool.Mode mode: Request mode (default is
+        PREFER_RW).
+        '''
+        return self._send(mode, 'eval', expr, *args)
+
+    def replace(self, space_name, values):
+        return self._send(Mode.RW, 'replace', space_name, values)
+
+    def insert(self, space_name, values):
+        return self._send(Mode.RW, 'insert', space_name, values)
+
+    def delete(self, space_name, key, **kwargs):
+        return self._send(Mode.RW, 'delete', space_name, key, **kwargs)
+
+    def upsert(self, space_name, tuple_value, op_list, **kwargs):
+        return self._send(Mode.RW, 'upsert', space_name, tuple_value,
+            op_list, **kwargs)
+
+    def update(self, space_name, key, op_list, **kwargs):
+        return self._send(Mode.RW, 'update', space_name, key, 
+            op_list, **kwargs)
+
+    def ping(self, mode=Mode.PREFER_RW, **kwargs):
+        '''
+        :param tarantool.Mode mode: Request mode (default is
+        PREFER_RW).
+        '''
+        return self._send(mode, 'ping', **kwargs)
+
+    def select(self, space_name, key, mode=Mode.PREFER_RO, **kwargs):
```

Comment:
There is https://peps.python.org/pep-3102/ regarding keyword-only arguments (it's sixteen birthday will be on this week, hooray!). It should work since Python 3.0.

(There is also https://peps.python.org/pep-0570/ regarding positional-only arguments. It should work since Python 3.8. It is irrelevant here, just interesting thing around.)

Even without PEP-3102 it is simple:

```python
def foo(**kwargs):
    x = kwargs.get('x')
    if x:
        del kwargs['x']
    <..do something on x..>
    return bar(**kwargs)
```

However PEP-3102 variant is more concise.

Reply:
Reworked the code to use `**kwargs` in methods were we already had `**kwargs`. Should I reword ones where there wasn't?
```python3
def insert(self, space_name, values, mode=Mode.RW):
    '''
    :param tarantool.Mode mode: Request mode (default is RW).
    '''

    return self._send(mode, 'insert', space_name, values)
```

Reply:
Why not use PEP-3102 syntax? For example current code for `select` is the following:

```python
def select(self, space_name, key, **kwargs):
    '''
    :param tarantool.Mode mode: Request mode (default is
    ANY).
    '''

    mode = kwargs.get('mode')
    if mode is None:
        mode = Mode.ANY
    else:
        del kwargs['mode']

    return self._send(mode, 'select', space_name, key, **kwargs)
```

It'll be the following:

```python
def select(self, space_name, key, *, mode=Mode.ANY, **kwargs):
    '''
    :param tarantool.Mode mode: Request mode (default is
    ANY).
    '''

    return self._send(mode, 'select', space_name, key, **kwargs)
```

I don't see any downsides in using this syntax in all connection interface methods.

Reply:
Implemented PEP-3102